### DISCUSSION TITLE
Unable to Connect to Waveshare RS485 to RJ45 to Inverters in Parallel

### DISCUSSION THREAD

**Discussion by gerd091986** [wills106/homeassistant-solax-modbus#604] [Q&A] [answered]
Hi All,

I recently thought I has another solution in using the Pocketwifi to control my Inverters but unfortunatly I have learned it's not stable enough.

I've been pulling my hair out trying to get this to work but have not been sucessfull, happy to buy coffee or make a donation for someones help with this. 

Home assistant is on 192.168.86.58 as a static ip on port 8123
Waveshare inverter is on: 192.168.86.59 port 502
Solax Wifi is on: 192.168.86.41 

- When I connect to the Solax Wifi dongle everything works fine, 256 entities and complete control

- When I am in Home assistant and set everything up, as soon as I click submit, the light goes blue on the waveshare device and it looks to connect, 23 seconds later the blue light goes out. in Viacom when the blue light comes on it states TCP connection is established, but Viacom status is not linked. So it looks like it is connecting but not properly, I get 1 device and 4 entities.

- There is no firewall to configure, my home assistant runs on a raspberry pi and home assistant OS is the operating system, correct me if I'm wrong

- Router - I have mapped the Waveshare port on 192.168.86.59 from 502 > 8123 so the home assistant can pick it up. Not sure if this was the right thing to do but Waveshare told me to make sure the ports are open etc

- Home Assistant - It's working correctly when connected to the Solax Pocket wifi, so it's configured correctly

- I have contacted SolaX to see if I need to enable anything and they have said there is nothing specific to setup or enable on the inverter

- I have restarted the inverters, Home Assistant and Waveshare device and it's still not working.

I have read through all the Wiki and discussions and I can't find a solution, any help would be greatly appreciated.

![Screenshot 2023-11-08 at 13 12 34](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/a2179917-4a53-4d8e-9293-3791fe9a629a)

![Screenshot 2023-11-08 at 13 14 00](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/a746dba8-27cf-4c71-97cf-9f81f711a527)

![Screenshot 2023-11-08 at 13 14 22](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/6eba4e96-9514-4a29-8bdc-e1fefabee142)

![Screenshot 2023-11-08 at 13 15 40](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/dc8302e2-3d96-445f-8020-cbc30851a386)

![PXL_20231108_131328223](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/d9b0203d-c7ce-4a11-a73b-49e738e4942f)

**Comment by gerd091986** [comment]
I have checked the logs after reading another post and it's definatly the connection to the Waveshare as it doesnt get a serial number as per [502](https://github.com/wills106/homeassistant-solax-modbus/discussions/520)

unrecognized inverter type - serial number : unknown
11:23:34 AM – (ERROR) SolaX Inverter Modbus (custom integration)
SolaX: cannot find serial number, even not for MIC
11:23:34 AM – (ERROR) SolaX Inverter Modbus (custom integration)
SolaX: reading serial number from address 0x0 failed; other address may succeed
11:23:34 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:23:29 AM and shows up 2 times

**Comment by gerd091986** [comment]
Bases on [this](https://community.home-assistant.io/t/universal-solar-inverter-over-modbus-rs485-tcp-custom-component-growatt-sofar-solax-solis/140143/756) thread I have tried to swap the pins.

Now I have the Red light on, green middle and last light Blue permanently, but now get more errors errors:

Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused
11:49:55 AM – (ERROR) SolaX Inverter Modbus (custom integration)
Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused
11:49:40 AM – (ERROR) SolaX Inverter Modbus (custom integration)
unrecognized inverter type - serial number : unknown
11:49:40 AM – (ERROR) SolaX Inverter Modbus (custom integration) - message first occurred at 11:23:34 AM and shows up 3 times
SolaX: cannot find serial number, even not for MIC
11:49:40 AM – (ERROR) SolaX Inverter Modbus (custom integration) - message first occurred at 11:23:34 AM and shows up 3 times
SolaX: reading serial number from address 0x0 failed; other address may succeed
11:49:40 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:23:29 AM and shows up 6 times
SolaX: attempt to read serialnumber failed at 0x0 data: None
11:49:40 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:49:40 AM and shows up 2 times
Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused
11:49:40 AM – (ERROR) SolaX Inverter Modbus (custom integration) - message first occurred at 11:49:40 AM and shows up 2 times

Swapped back the pins as I don't think this helped.

**Comment by cleveHEX** [comment]
I don't have any personal experience with this device, but can you set Modbus Gateway Type to "Simple Modbus to TCP/IP mode"? 
I might be very off then pardon me, but it does not match imho:
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/9021297/cc322b1a-a376-4b42-a2ce-c52177a8a1ca)
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/9021297/31b4908f-6d35-413a-8ff4-da42e6472700)

**Reply by gerd091986** [reply]
No need for a pardon, will take all the help I can get. Thank you!

when I go through Viacom it's set to Simple Modbus to TCP to RTU mode:
![WhatsApp Image 2023-11-09 at 12 02 10](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/d4ed3883-8185-4357-8185-25eadaabbba7)

When I go via the web interface it's set to Modbus TCP to RTU:
![Screenshot 2023-11-09 at 12 05 15](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/1bea7fae-3e09-460c-920f-5521268c8060)

**Reply by cleveHEX** [reply]
I source the info from the repo wiki, what are the other options (screenshot would be nice)? Is there an exact option to select as I provided?

**Reply by gerd091986** [reply]
No exact option as you described.

On this web interface:
![WhatsApp Image 2023-11-09 at 12 16 18](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/d4ec419f-462b-410e-b841-c5689b353e84)

On Viacom:
![WhatsApp Image 2023-11-09 at 12 14 53](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/6038373b-b425-41d0-aa99-3264a05157cf)

**Reply by gerd091986** [reply]
The other option in Viacom if this helps:

![WhatsApp Image 2023-11-09 at 12 19 17](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/ae5d86ea-10a6-4806-90de-b73e26807eb5)

**Reply by cleveHEX** [reply]
My last guess is to try 
![obrazek](https://github.com/wills106/homeassistant-solax-modbus/assets/9021297/5336617e-f512-4490-a099-622fd5921bb9)
when setting up the integration. If it does not help, someone else will have to step in, as I have ran out of ideas.

**Reply by gerd091986** [reply]
Tried that again and still, I get the blue light for 30sec and then goes green. 

Thank you for all your help trying @cewbdex I really appreciate it.

I will keep reading and trying, I'm sure I will get there in the end.

**Comment by gerd091986** [comment]
Also I am recieving the short Green ACT periods mentioned [here](https://community.home-assistant.io/t/universal-solar-inverter-over-modbus-rs485-tcp-custom-component-growatt-sofar-solax-solis/140143/772?u=gerd091986)

**Comment by gerd091986** [comment]


**Reply by cleveHEX** [reply]
The barcode contains your serial number and edit history is public, you need to delete this post if you want your inverter SN remain private.

**Reply by gerd091986** [reply]
Thank you.

Lets try that again:

![WhatsApp Image 2023-11-09 at 12 53 57](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/87a17698-30c9-4346-ba19-93f41226c4e0)

**Comment by gerd091986** [comment]
Waveshare suggested updating the firmware of the device, here is the latest firmware if anyone needs it:

[1.490 firmware](https://files.waveshare.com/wiki/Serial-TO-ETH-B/WSDEV-1.490.BIN)

I updated the firmware on the Waveshare device, turned the inverters completely off, waited 10min and then turned them back on.

Still when connecting via Home Assistant I only get 4 devices and the blue light for 30 seconds which then goes off and doesnt come back on. 

If anyone has any further ideas I'm all ears, otherwise I will keep digging.

**Comment by wills106** [comment]
> * Router - I have mapped the Waveshare port on 192.168.86.59 from 502 > 8123 so the home assistant can pick it up. Not sure if this was the right thing to do but Waveshare told me to make sure the ports are open etc



> Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused
> 11:49:55 AM – (ERROR) SolaX Inverter Modbus (custom integration)


That combination won't work. If you have managed to route your port 502 on the waveshare to 8123 and you are pointing the Integration at 502 that won't match up.

I would remove any firewall rules you have setup and start from the beginning.

Let the Integration point to port 502.
Might be worth doing a factory reset on the Waveshare.
Modbus Gateway Type to "Simple Modbus to TCP/IP mode"
Integration connects with "The Modbus TCP variant" as "Modbus TCP"
Do you actually have more than one Inverter connected together? If you only have one there is no point in setting "Parallel Mode (Master-Slave)"

**Reply by gerd091986** [reply]
Thank you for coming back to me.

> That combination won't work. If you have managed to route your port 502 on the waveshare to 8123 and you are pointing the Integration at 502 that won't match up.

I removed this previously as I realised it wasnt right.

> I would remove any firewall rules you have setup and start from the beginning.

I have removed everything except for a static IP for the Waveshare device and Home Assistant, in the same way I have for the Pocketwifi that is working.

I have also tried it with DHCP and removing the Static IP rules, still no luck.

> Let the Integration point to port 502.

I have removed all port rules.

> Might be worth doing a factory reset on the Waveshare.

- Checked all the connections to the device
- Changed the Ethernet cables incase there was something wrong with them
- Factory reset the Waveshare by connecting the two pins and turning on
- Tried to connect to it after factory reset, light turns blue then goes back to green 30sec later, 4 entities, same as before
- Loaded the latest firmware
- Connected again, light turns blue then goes back to green 30sec later, 4 entities same as before

> Modbus Gateway Type to "Simple Modbus to TCP/IP mode"
> Integration connects with "The Modbus TCP variant" as "Modbus TCP"

I don't have this option, I have 'Simple Modbus TCP to RTU', as per the screenshots in the wiki.

> Do you actually have more than one Inverter connected together? If you only have one there is no point in setting "Parallel Mode (Master-Slave)"

I have 2 inverters in Master and Slave mode, I have the ethernet cable connected to the Master.

**Reply by gerd091986** [reply]
Just looking at this error in more detail:

> This error originated from a custom integration.
> 
> Logger: pymodbus.logging
> Source: custom_components/solax_modbus/__init__.py:350
> Integration: SolaX Inverter Modbus (documentation, issues)
> First occurred: November 9, 2023 at 11:49:55 AM (32 occurrences)
> Last logged: 10:14:39 AM
> 
> Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused
> Unable to decode response Modbus Error: Unknown response 57
> Modbus IO exception Modbus Error: [Input/Output] Unable to decode request
> Connection to (192.168.86.59, 502) failed: [Errno 113] Host is unreachable
> Connection to (192.168.86.59, 502) failed: timed out

> Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused

Thats my wifi router IP address that's refusing the connection, will look into why this might happen more.

**Reply by wills106** [reply]
I didn't notice that, I was just focusing on the port number.

**Reply by gerd091986** [reply]
Added Port Forwarding for the Waveshare Device 502 >502, this error is gone:

> Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused

Left with:

>  unrecognized inverter type - serial number : unknown
> 11:05:22 AM – (ERROR) SolaX Inverter Modbus (custom integration)
> SolaX: cannot find serial number, even not for MIC
> 11:05:22 AM – (ERROR) SolaX Inverter Modbus (custom integration)
> SolaX: reading serial number from address 0x0 failed; other address may succeed
> 11:05:22 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:05:17 AM and shows up 2 times

Connection still not being established.

Removed port forwarding for Waveshare device and added it for 502>502 for Home Assistant (Because I have no idea what I'm doing and trying everything)this error is gone:

> Connection to (192.168.86.1, 502) failed: [Errno 111] Connection refused

I'm left with:

unrecognized inverter type - serial number : unknown
11:11:14 AM – (ERROR) SolaX Inverter Modbus (custom integration)
SolaX: cannot find serial number, even not for MIC
11:11:14 AM – (ERROR) SolaX Inverter Modbus (custom integration)
SolaX: reading serial number from address 0x0 failed; other address may succeed
11:11:14 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:11:09 AM and shows up 2 times

Removed the Port Forwarding rules an that error didnt come up again, only got these errors:

> unrecognized inverter type - serial number : unknown
> 11:13:26 AM – (ERROR) SolaX Inverter Modbus (custom integration)
> SolaX: cannot find serial number, even not for MIC
> 11:13:26 AM – (ERROR) SolaX Inverter Modbus (custom integration)
> SolaX: reading serial number from address 0x0 failed; other address may succeed
> 11:13:26 AM – (WARNING) SolaX Inverter Modbus (custom integration) - message first occurred at 11:13:21 AM and shows up 2 times 

Seems like a red herring, will stop going down this path.

**Reply by wills106** [reply]
Just want to double check you are connecting your Waveshare to the COM Port RS485 and not the Modbus Power Meter RS485 or the Inverter link up RS485?

**Reply by gerd091986** [reply]
I think I'm connected to the right port, the Com/LCD port.

![PXL_20231110_115539709.jpg](https://github.com/wills106/homeassistant-solax-modbus/assets/149315629/2a7266fb-9fc8-4696-bf54-d6b9b2fdb4d6)

**Comment by gerd091986** [comment]
Disconnected the Slave battery and all 256 entities came up.

Something to do with connecting a slave then, now time to reconenct the slave without losing the connection.

**Comment by gerd091986** [comment]
1. Changed Main Inverter to be Master and connection stayed.
2. Connected communication cable between Master and Slave
3. Turned on Slave - Lost everything

**Reply by cleveHEX** [reply]
Is fw version same across your batteries?

**Reply by gerd091986** [reply]
Yes it is, I called Solax and they checked that the batteries are on the same firmware etc. which they are.

They then said, they dont think Modbus is supported in parallel, I told them it's working on Pocket Wifi, they said they will send it over to the R&D department for more infomation. 

I'm interested if anyone with a Parallel setup has managed to get this to work....

**Reply by gerd091986** [reply]
Looks like I'm out of luck....


>  Dear customer,
> 
> I just got a reply from the R&D department. They say that Modbus connection is not an option when the inverter is in parallel mode. Sorry for the waiting time and the inconvenience. 
> 
> Best Regards,

**Reply by cleveHEX** [reply]
Interesting reply. I know one guy that had parallel-slave setup and he contacted me that he uses modbus adapter on master and he does not see most entities from the slave and we concluded that the information is not passed from slave to master and he needs to connect the second inverter too. But we weren't in contact since and I don't know the results. I don't remember which inverter he used, only that it was probably G4.

**Reply by gerd091986** [reply, marked as answer]
They were lying....

Just change the Modbus address from 1 to 2 on the slave. Big thanks to @fxstein for this [comment](https://github.com/wills106/homeassistant-solax-modbus/discussions/618#discussioncomment-7555254) that got me there.

I will stop annoying everyone now!

**Reply by cleveHEX** [reply]
How could it work before in parallel mode, when they had the same addresses? I hope you installed the inverters yourself and it was not some company who did the mistake.

**Reply by gerd091986** [reply]
Parallel is done over a seperate hardwired connection cable, no options there, it sorts itself out, just one setting to change on each inverter. Has been working fine for a year and it doesnt say anywhere you need to change the Modbus address.

Modbus has it's connection through the COM/LCD port and it's own settings, one of which is the address. I should have thought sooner to change the address from 1 to 2 on the slave so they are not the same.

**Reply by cleveHEX** [reply]
So it's your work? If so, it's fine, SolaX does not really provide proper public documentation IMHO, but a company has to know.
Made <https://github.com/wills106/homeassistant-solax-modbus/pull/626>, so it could help people like you in future.

**Reply by gerd091986** [reply]
Yes my work, a company installed the system a year ago in parallel but did not connect the modbus.

I'm doing the Modbus connection and setup myself.

Thanks for adding that to the FAQ!

**Reply by meb31** [reply]
I lost most of my hair over the past two weeks trying exactly the same as you! It was so frustrating and time consuming checking everything and trying everything then I saw the post you spotted and it works. Such relief, but I wish it had been on the faq.

**Reply by cleveHEX** [reply]
It's in docs faq since we solved it.
